### PR TITLE
linter: add finally return check for unreachable checker

### PR DESCRIPTION
**Pull request by i582**
Report the use of `return` in `finally` when
there are other `return`, `throw` in the `try-catch`.

The problem is they will be unreachable.

For example:

```php
function f(): int {
  try {
    throwException();
  } catch (Exception $_) {
    return 2; // unreachable
  } finally {
    return 1;
  }
}

f(); // always return 1
```

Report the use of `die` in `catch` when
there are `finally` block.

The problem is `finally` will be unreachable.

```php
function f(): int {
    try {
        throwException();
    } catch (RuntimeException $_) {
        return 2;
    } catch (Exception $_) {
        die();
    } finally { // block finally is unreachable (because 2 catch block contains a exit/die)
        return 1;
    }
}
```

#332

### CODE REVIEW

File: src/linter/block.go [hunk 1]
```diff
@@ -723,16 +726,112 @@ func (b *blockWalker) handleTry(s *ir.TryStmt) bool {
...
 }
 
+func (b *blockWalker) checkUnreachableForFinallyReturn(tryStmt *ir.TryStmt, tryCtx *blockContext, finallyCtx *blockContext, catchContexts []*blockContext) {
+	if finallyCtx == nil {
+		return
+	}
+
+	var exitPoints []exitPoint
+
+	// if try contains some other return/die/etc statement
```
Comment (quasilyte) [inline src/linter/block.go:746]: `s/if/If` + period in the end.

File: src/linter/block.go [hunk 2]
```diff
@@ -723,16 +726,112 @@ func (b *blockWalker) handleTry(s *ir.TryStmt) bool {
...
+}
+
+type exitPoint struct {
+	n   ir.Node
+	typ string
+}
+
+func (b *blockWalker) findExitPointsByFlags(where ir.Node, exitFlags int) (points []exitPoint) {
+	irutil.Inspect(where, func(n ir.Node) bool {
+		if exitFlags&FlagReturn == FlagReturn {
```
Comment (quasilyte) [inline src/linter/block.go:795]: Do you want to check that this bit is set?
Why not do it like this:
```go
exitFlags&FlagReturn != 0
```
It's less error-prone after the copy/paste.

File: src/linter/block.go (hunk 1 above)
Comment (i582) [reply src/linter/block.go]: Fixed.

File: src/tests/inline/testdata/checkers/finallyReturn.php [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        return 1;
+    }
+}
+
+function finallyReturnOkWithoutReturnInTryCatch(): int {
+    try {
+        throwException();
+    } catch (Exception $_) {
+    } finally {
+        return 1; /* ok, catch and try blocks don't contain return/exceptions/die/etc */
```
Comment (quasilyte) [inline src/tests/inline/testdata/checkers/finallyReturn.php:106]: No reason to use `/**` comment here.

Comment (i582) [reply src/tests/inline/testdata/checkers/finallyReturn.php]: Fixed.

File: src/linter/block.go (hunk 2 above)
Comment (i582) [reply src/linter/block.go]: Fixed.

**Review by quasilyte — APPROVED** [review]

**Pull request merged**